src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/CmsSignature.cs

```diff
@@ -83,7 +92,21 @@ internal static CmsSignature Resolve(string signatureAlgorithmOid)
             }
 
             byte[] signature;
-            bool signed = processor.Sign(dataHash, hashAlgorithmName, certificate, silent, out oid, out signature);
+            bool signed = processor.Sign(dataHash, hashAlgorithmName, certificate, key, silent, out oid, out signature);
+
+            if (signed && key != null)
+            {
+                // signatureParameters is used to figure out padding for RSA and ignored for ECDsa and DSA
+                // For RSA we can only get Pkcs1 as Pss does not support signing and null will default to Pkcs1 on RSAPkcs1CmsSignature instance.
+                if (!processor.VerifySignature(dataHash, signature, oid.Value, hashAlgorithmName, signatureParameters: null, certificate))
```

Comment:
It seems weird to add this with a comment justifying that this pattern only works because of a gap in the API.

It feels like a lot of re-parsing would be saved by just doing the check inside the sign method where the verification parameters are the same as the signature parameters...

Reply:
sounds good, will update